### PR TITLE
Update so that all moves are on board

### PR DESCRIPTION
**Pull request by huntedhippo**
To do this I have simplified the get_moves functions by just calling validate_move.

Validate move checks that they're onboard.

### CODE REVIEW

File: server/chess.py [hunk 1]
```diff
@@ -77,6 +77,15 @@ def path_is_empty(
         victim = self.get_piece(rank, file)
         return victim.side != piece.side
 
+    def on_board(rank: int, file: int) -> bool:
+        """check if valid square i.e. rank and file on board, rename to valid_square if you want"""
+        """move min/max_rank and min/max_file to other file if appropriate"""
```
Comment (Artemis21) [inline server/chess.py]:
```suggestion
    def on_board(self, rank: int, file: int) -> bool:
        """Check if valid square i.e. rank and file on board."""
```

File: server/chess.py [hunk 2]
```diff
@@ -77,6 +77,15 @@ def path_is_empty(
...
+        min_rank = 0
+        max_rank = 7
+        min_file = 0
+        max_file = 7
+        return rank >= min_rank and rank <= max_rank and file >= min_file and file <= max_file
```
Comment (Artemis21) [inline server/chess.py]:
```suggestion
        return (
            rank >= min_rank and rank <= max_rank
            and file >= min_file and file <= max_file
        )
```

File: server/chess.py [hunk 3]
```diff
@@ -85,14 +94,10 @@ def get_moves_in_direction(
         while True:
             file += file_direction
             rank += rank_direction
-            target = self.get_piece(rank, file)
-            if target:
-                if target.side == piece.side:
-                    break
+            if validate_move(self, piece.rank, piece.file, rank, file):
                 yield rank, file
+            else
```
Comment (Artemis21) [inline server/chess.py]:
```suggestion
            else:
```

File: server/chess.py [hunk 4]
```diff
@@ -152,28 +157,9 @@ def validate_pawn_move(
...
-                en_passant_valid = (
-                    en_passant_pawn and en_passant_pawn.side != pawn.side
-                    and en_passant_pawn.first_move_last_turn
-                )
-                if en_passant_valid:
-                    yield rank, file
-            if target and target.side != pawn.side:
+        for rank_absolute, file in ((1, 0), (2, 0), (1, -1), (1, 1)):
+            rank = rank_absolute * pawn.side.forwards
+            if validate_move(self, pawn.rank, pawn.file rank, file):
```
Comment (Artemis21) [inline server/chess.py]:
```suggestion
        options = ((1, 0), (2, 0), (1, -1), (1, 1))
        for absolute_rank_delta, file_delta in options:
            rank = pawn.rank + absolute_rank_delta * pawn.side.forwards
            file = pawn.file + file_delta
            if validate_move(self, pawn.rank, pawn.file, rank, file):
```

File: server/chess.py [hunk 5]
```diff
@@ -209,8 +195,7 @@ def get_knight_moves(
                 for file_direction in (-1, 1):
                     rank = knight.rank + rank_absolute * rank_direction
                     file = knight.file + file_absolute * file_direction
-                    victim = self.get_piece(rank, file)
-                    if (not victim) or (victim.side != knight.side):
+                    if validate_move(self, knight.rank, knight.file, rank, file):
```
Comment (Artemis21) [inline server/chess.py]: This method is slower - it re-fetches the piece from the database, and performs a whole lot of checks we have already done, for very little gain in code complexity.

File: server/chess.py [hunk 6]
```diff
@@ -318,8 +290,7 @@ def get_king_moves(self, king: models.Piece) -> typing.Iterator[int, int]:
                     continue
                 rank = king.rank + rank_direction
                 file = king.file + file_direction
-                victim = self.get_piece(rank, file)
-                if (not victim) or (victim.side != king.side):
+                if validate_move(self, king.rank, king.file, rank, file):
```
Comment (Artemis21) [inline server/chess.py]: See comment on get_knight_moves

File: server/chess.py [hunk 7]
```diff
@@ -290,22 +275,9 @@ def get_allowed_castling(self, king: models.Piece) -> typing.List[
...
-                    empty_squares_empty = False
-                    break
-            if not empty_squares_empty:
-                continue
-            valid = not self.hypothetical_check(
-                king.side, (king, king.rank, king_end),
-                (rook, king.rank, rook_end)
-            )
-            if valid:
+        for king_end, rook_start in ((2,0),(6,7)):
```
Comment (Artemis21) [inline server/chess.py]:
```suggestion
        for king_end, rook_start in ((2, 0), (6, 7)):
```

**Review by Artemis21 — CHANGES_REQUESTED** [review]
I reluctantly accept this method because it does make the code a lot less complex in a couple of cases. I had considered and rejected it before due to it's inefficiency though.

File: server/chess.py (hunk 1 above)
Comment (huntedhippo) [reply server/chess.py]: I actually didn't use this method in the end, as I switched to just using validate move.  Shall I delete?

File: server/chess.py (hunk 5 above)
Comment (huntedhippo) [reply server/chess.py]: For starters it makes sure the knight doesn't jump off the board.

Comment (huntedhippo) [reply server/chess.py]: Secondly it ensures the knight doesn't move to put his king in check

**Comment by huntedhippo** [comment]
I'm under the impression that efficiency is not the number one priority with the implementation of this code, and anyway, the validate move barely decreases efficiency, but does remove redundant code.

I've added validate_move_piece to address your concerns about the efficiency of additional calls to the database. Particularly in the get_moves functions this could slow down the process a lot.

I think it is efficient to check that the moves are on_board and not causing check though when you check moves so that it only returns valid moves. Otherwise you'll have to validate them afterwards.  Further, it stops you from perpetually generating moves off board in your get_moves_in_direction function

File: server/chess.py (hunk 5 above)
Comment (Artemis21) [reply server/chess.py]: The latter is already ensured on line 371, the former should be checked somewhere similar.

File: server/chess.py (hunk 1 above)
Comment (Artemis21) [reply server/chess.py]: Sure, can delete the function.

File: server/chess.py [hunk 8]
```diff
@@ -77,6 +77,17 @@ def path_is_empty(
...
+        """check if valid square i.e. rank and file on board"""
+        min_rank = 0
+        max_rank = 7
+        min_file = 0
+        max_file = 7
+        return (
+            rank >= min_rank and rank <= max_rank
+            and file >= min_file and file <= max_file
+        )
+
```
Comment (Artemis21) [inline server/chess.py:90]:
```suggestion
```
Not used.

File: server/chess.py [hunk 9]
```diff
@@ -85,14 +96,10 @@ def get_moves_in_direction(
...
-            target = self.get_piece(rank, file)
-            if target:
-                if target.side == piece.side:
-                    break
+            if validate_move_piece(self, piece, rank, file):
                 yield rank, file
-                break
             else:
-                yield rank, file
+                break
```
Comment (Artemis21) [inline server/chess.py:102]: This does not cover the case where a move would be invalid but travelling further would be valid (eg. travelling to protect the king)

File: server/chess.py [hunk 10]
```diff
@@ -85,14 +96,10 @@ def get_moves_in_direction(
         while True:
             file += file_direction
             rank += rank_direction
-            target = self.get_piece(rank, file)
-            if target:
-                if target.side == piece.side:
-                    break
+            if validate_move_piece(self, piece, rank, file):
                 yield rank, file
```
Comment (Artemis21) [inline server/chess.py:100]: The only check that need be added here is that the rank and file are in range, this seems inefficient.

File: server/chess.py [hunk 11]
```diff
@@ -152,28 +159,11 @@ def validate_pawn_move(
...
-                    and en_passant_pawn.first_move_last_turn
-                )
-                if en_passant_valid:
-                    yield rank, file
-            if target and target.side != pawn.side:
+        options = ((1, 0), (2, 0), (1, -1), (1, 1))
+        for absolute_rank_delta, file_delta in options:
+            rank = pawn.rank + absolute_rank_delta * pawn.side.forwards
+            file = pawn.file + file_delta
+            if validate_move_piece(self, pawn, rank, file):
```
Comment (Artemis21) [inline server/chess.py:166]:
```suggestion
            if self.validate_move_piece(pawn, rank, file):
```
Cause that's how python works :wink:

File: server/chess.py [hunk 12]
```diff
@@ -209,8 +199,7 @@ def get_knight_moves(
                 for file_direction in (-1, 1):
                     rank = knight.rank + rank_absolute * rank_direction
                     file = knight.file + file_absolute * file_direction
-                    victim = self.get_piece(rank, file)
-                    if (not victim) or (victim.side != knight.side):
+                    if validate_move_piece(self, knight, rank, file):
```
Comment (Artemis21) [inline server/chess.py:202]:
```suggestion
                    if self.validate_move_piece(knight, rank, file):
```
If you must use this method, but it stills seems needlessly inefficient.

File: server/chess.py [hunk 13]
```diff
@@ -288,24 +277,10 @@ def get_allowed_castling(self, king: models.Piece) -> typing.List[
...
-                    empty_squares_empty = False
-                    break
-            if not empty_squares_empty:
-                continue
-            valid = not self.hypothetical_check(
-                king.side, (king, king.rank, king_end),
-                (rook, king.rank, rook_end)
-            )
-            if valid:
+        for king_end, rook_start in ((2, 0),(6, 7)):
```
Comment (Artemis21) [inline server/chess.py:281]:
```suggestion
        for king_end, rook_start in ((2, 0), (6, 7)):
```

File: server/chess.py [hunk 14]
```diff
@@ -288,24 +277,10 @@ def get_allowed_castling(self, king: models.Piece) -> typing.List[
...
+            if validate_move_piece(self, king, king.rank, king_end):
```
Comment (Artemis21) [inline server/chess.py:282]:
```suggestion
            if self.validate_move_piece(king, king.rank, king_end):
```

File: server/chess.py [hunk 15]
```diff
@@ -318,8 +293,7 @@ def get_king_moves(self, king: models.Piece) -> typing.Iterator[int, int]:
                     continue
                 rank = king.rank + rank_direction
                 file = king.file + file_direction
-                victim = self.get_piece(rank, file)
-                if (not victim) or (victim.side != king.side):
+                if validate_move_piece(self, king, rank, file):
```
Comment (Artemis21) [inline server/chess.py:296]:
```suggestion
                if self.validate_move_piece(king, rank, file):
```
> If you must use this method, but it stills seems needlessly inefficient.

File: server/chess.py [hunk 16]
```diff
@@ -328,17 +302,23 @@ def validate_move(
             self, start_rank: int, start_file: int, end_rank: int,
             end_file: int, check_allowed: bool = False) -> bool:
         """Validate a move."""
-        if start_file == end_file and start_rank == end_rank:
+        piece = self.get_piece(start_rank, start_file)
+        if not piece:
+            return False
+        return validate_move_piece(self, piece, end_rank, end_file, check_allowed)
```
Comment (Artemis21) [inline server/chess.py:308]:
```suggestion
        return self.validate_move_piece(piece, end_rank, end_file, check_allowed)
```

**Review by Artemis21 — CHANGES_REQUESTED** [review]
Please lint before pushing!

**Comment by huntedhippo** [comment]
Closed this as deleted repository by accident

**Pull request closed without merging**